fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/remote/LAONetworkManager.java

```diff
@@ -0,0 +1,194 @@
+import com.github.dedis.popstellar.model.network.method.Publish;
+import com.github.dedis.popstellar.model.network.method.Query;
+import com.github.dedis.popstellar.model.network.method.Subscribe;
+import com.github.dedis.popstellar.model.network.method.Unsubscribe;
+import com.github.dedis.popstellar.model.network.method.message.MessageGeneral;
+import com.github.dedis.popstellar.model.network.method.message.data.Data;
+import com.github.dedis.popstellar.model.objects.security.KeyPair;
+import com.github.dedis.popstellar.repository.LAORepository;
+import com.github.dedis.popstellar.utility.error.DataHandlingException;
+import com.github.dedis.popstellar.utility.error.JsonRPCErrorException;
+import com.github.dedis.popstellar.utility.handler.MessageHandler;
+import com.github.dedis.popstellar.utility.scheduler.SchedulerProvider;
+import com.google.gson.Gson;
+import com.tinder.scarlet.WebSocket;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.reactivex.Completable;
+import io.reactivex.Observable;
+import io.reactivex.Single;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.subjects.PublishSubject;
+import io.reactivex.subjects.Subject;
+
+/** This class handles the JSON-RPC layer of the protocol */
+public class LAONetworkManager implements MessageSender {
+
+  private static final String TAG = LAONetworkManager.class.getSimpleName();
+
+  private final LAORepository repository;
+  private final MessageHandler messageHandler;
+  private final Connection connection;
+  public final AtomicInteger requestCounter = new AtomicInteger();
+  private final SchedulerProvider schedulerProvider;
+  private final Gson gson;
+
+  // A subject that represents unprocessed messages
+  private final Subject<GenericMessage> unprocessed = PublishSubject.create();
+  private final List<String> subscribedChannels = new LinkedList<>();
+  private final CompositeDisposable disposables;
+
+  public LAONetworkManager(
+      LAORepository repository,
+      MessageHandler messageHandler,
+      Connection connection,
+      Gson gson,
+      SchedulerProvider schedulerProvider) {
+    this.repository = repository;
+    this.messageHandler = messageHandler;
+    this.connection = connection;
+    this.gson = gson;
+    this.schedulerProvider = schedulerProvider;
+    this.disposables = new CompositeDisposable(subscribeToUpstream(), subscribeToWebsocketEvents());
+  }
+
+  private Disposable subscribeToWebsocketEvents() {
+    return connection
+        .observeWebsocket()
+        .subscribeOn(schedulerProvider.io())
+        .filter(event -> event.getClass().equals(WebSocket.Event.OnConnectionOpened.class))
+        .subscribe(
+            event ->
+                subscribedChannels.forEach(
+                    channel -> disposables.add(subscribe(channel).subscribe())));
+  }
```

Comment:
The intention is not clear here.
If I understand correctly, what we're really doing, is not subscribing to websocket events.
We're ensuring that we re-subscribe to channels on reconnection.
How about a more explicit function name?

I would add a few comments within the function as well, to document what each line is doing behind the magic of it all.